pkg/controller/elasticsearch/driver/upscale.go

```diff
@@ -41,25 +41,28 @@ func HandleUpscaleAndSpecChanges(
 	ctx upscaleCtx,
 	actualStatefulSets sset.StatefulSetList,
 	expectedResources nodespec.ResourcesList,
-) error {
+) (sset.StatefulSetList, error) {
 	// adjust expected replicas to control nodes creation and deletion
 	adjusted, err := adjustResources(ctx, actualStatefulSets, expectedResources)
 	if err != nil {
-		return err
+		return nil, err
 	}
 	// reconcile all resources
 	for _, res := range adjusted {
 		if err := settings.ReconcileConfig(ctx.k8sClient, ctx.es, res.StatefulSet.Name, res.Config); err != nil {
-			return err
+			return nil, err
 		}
 		if _, err := common.ReconcileService(ctx.k8sClient, ctx.scheme, &res.HeadlessService, &ctx.es); err != nil {
-			return err
+			return nil, err
 		}
-		if err := sset.ReconcileStatefulSet(ctx.k8sClient, ctx.scheme, ctx.es, res.StatefulSet); err != nil {
-			return err
+		reconciled, err := sset.ReconcileStatefulSet(ctx.k8sClient, ctx.scheme, ctx.es, res.StatefulSet)
+		if err != nil {
+			return nil, err
 		}
+		// update actual with the reconciled ones for newt steps to work with up-to-date information
+		actualStatefulSets = actualStatefulSets.WithStatefulSet(reconciled)
 	}
-	return nil
+	return actualStatefulSets, nil
```

Comment:
Would returning `adjusted` be more clear here?

Reply:
You mean in order to include config and service?
One tricky detail is that `adjusted` contains only StatefulSet we expect, for which we do have operations to perform.
StatefulSets in the process of being scaled down exist in `actualStatefulSets` but not in `adjusted`.

Reply:
> StatefulSets in the process of being scaled down exist in actualStatefulSets but not in adjusted.

Here, by scaled down you mean specifically scaled down to zero, ie. being removed? If yes, then I get it. If not, I almost get it:), but it seems to me that if sset is in `expected`, it will be in `actualStatefulSets`.

Reply:
Yes, sorry, I meant being removed.

Reply:
Got it, thanks.